### ISSUE TITLE
ZeroTrust - The balance.unsettledAmount is missing in the calculations for  `getMaxWithdraw`  and `isSubAmountAllowed` in UsdPool.sol

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-05-elfi-protocol-judging#236]
ZeroTrust

Medium

# The balance.unsettledAmount is missing in the calculations for  `getMaxWithdraw`  and `isSubAmountAllowed` in UsdPool.sol

## Summary
The balance.unsettledAmount is missing in the calculations for  `getMaxWithdraw`  and `isSubAmountAllowed` in UsdPool.sol
## Vulnerability Detail
```javascript
    function getMaxWithdraw(Props storage self, address stableToken) public view returns (uint256) {
        TokenBalance storage balance = self.stableTokenBalances[stableToken];
        uint256 poolLiquidityLimit = getPoolLiquidityLimit();
        if (poolLiquidityLimit == 0) {
@>>            return balance.amount - balance.holdAmount;
        } else {
            uint256 holdNeedAmount = CalUtils.divRate(balance.holdAmount, poolLiquidityLimit);
@>>            return balance.amount > holdNeedAmount ? balance.amount - holdNeedAmount : 0;
        }
    }
```

```javascript
    function isSubAmountAllowed(Props storage self, address stableToken, uint256 amount) public view returns (bool) {
        TokenBalance storage balance = self.stableTokenBalances[stableToken];
        if (balance.amount < amount) {
            return false;
        }
        uint256 poolLiquidityLimit = getPoolLiquidityLimit();
        if (poolLiquidityLimit == 0) {
@>>            return balance.amount - balance.holdAmount >= amount;
        } else {
@>>            return CalUtils.mulRate(balance.amount - amount, poolLiquidityLimit) >= balance.holdAmount;
        }
    }

```
We can see that the balance.unsettledAmount is missing in the calculations. The balance.unsettledAmount represents the fees earned by the pool, but the assets have not yet been transferred. 
## Impact
The higher-level function calls to getMaxWithdraw and isSubAmountAllowed should return true, but they return false instead, preventing the function from continuing to execute correctly.

## Code Snippet
https://github.com/sherlock-audit/2024-05-elfi-protocol/blob/main/elfi-perp-contracts/contracts/storage/UsdPool.sol#L214

https://github.com/sherlock-audit/2024-05-elfi-protocol/blob/main/elfi-perp-contracts/contracts/storage/UsdPool.sol#L241

## Tool used

Manual Review

## Recommendation
```diff
    function getMaxWithdraw(Props storage self, address stableToken) public view returns (uint256) {
        TokenBalance storage balance = self.stableTokenBalances[stableToken];
        uint256 poolLiquidityLimit = getPoolLiquidityLimit();
        if (poolLiquidityLimit == 0) {
-            return balance.amount - balance.holdAmount;
+               return balance.amount + balance.unsettledAmount - balance.holdAmount;
        } else {
            uint256 holdNeedAmount = CalUtils.divRate(balance.holdAmount, poolLiquidityLimit);
-            return balance.amount > holdNeedAmount ? balance.amount - holdNeedAmount : 0;
+            return balance.amount + balance.unsettledAmount > holdNeedAmount ? balance.amount + balance.unsettledAmount - holdNeedAmount : 0;
        }
    }
```

```diff

    function isSubAmountAllowed(Props storage self, address stableToken, uint256 amount) public view returns (bool) {
        TokenBalance storage balance = self.stableTokenBalances[stableToken];
        if (balance.amount < amount) {
            return false;
        }
        uint256 poolLiquidityLimit = getPoolLiquidityLimit();
        if (poolLiquidityLimit == 0) {
-            return balance.amount - balance.holdAmount >= amount;
+            return balance.amount + balance.unsettledAmount - balance.holdAmount >= amount;
        } else {
-            return CalUtils.mulRate(balance.amount - amount, poolLiquidityLimit) >= balance.holdAmount;
+            return CalUtils.mulRate(balance.amount + balance.unsettledAmount - amount, poolLiquidityLimit) >= balance.holdAmount;
        }
    }

```

**Comment by 0502lian** [comment]
Escalate
This issue should be high.
There is a calculation error inside the `UsdPool::isSubAmountAllowed()` function.
`UsdPool::isSubAmountAllowed()` is called by `UsdPool::subStableToken()`
```javascript
   function subStableToken(Props storage self, address stableToken, uint amount) external {
@>>        require(isSubAmountAllowed(self, stableToken, amount), "sub failed with balance not enough");
        UsdPoolTokenUpdateCache memory cache = _convertBalanceToCache(
            stableToken,
            self.stableTokenBalances[stableToken]
        );
        self.stableTokenBalances[stableToken].amount -= amount;
        cache.amount = self.stableTokenBalances[stableToken].amount;
        _emitPoolUpdateEvent(cache);
    }
```

The functions `ClaimRewardsProcess.claimRewards()`, `LpPoolProcess.updatePnlAndUnHoldPoolAmount()` (called by decreasePosition), and `RedeemProcess._redeemStakeUsd()` internally call `UsdPool::subStableToken()`, then will revert,    which causes users to be unable to claim rewards, reduce positions, and redeem StakeUsd. Additionally, if no new funds are added by other users, the impact could last for over a week.
According to [Sherlock’s rules](https://docs.sherlock.xyz/audits/judging/judging#v.-how-to-identify-a-medium-issue), “The issue causes locking of funds for users for more than a week,” it qualifies as High.

**Comment by sherlock-admin3** [comment]
> Escalate
> This issue should be high.
> There is a calculation error inside the `UsdPool::isSubAmountAllowed()` function.
> `UsdPool::isSubAmountAllowed()` is called by `UsdPool::subStableToken()`
> ```javascript
>    function subStableToken(Props storage self, address stableToken, uint amount) external {
> @>>        require(isSubAmountAllowed(self, stableToken, amount), "sub failed with balance not enough");
>         UsdPoolTokenUpdateCache memory cache = _convertBalanceToCache(
>             stableToken,
>             self.stableTokenBalances[stableToken]
>         );
>         self.stableTokenBalances[stableToken].amount -= amount;
>         cache.amount = self.stableTokenBalances[stableToken].amount;
>         _emitPoolUpdateEvent(cache);
>     }
> ```
> 
> The functions `ClaimRewardsProcess.claimRewards()`, `LpPoolProcess.updatePnlAndUnHoldPoolAmount()` (called by decreasePosition), and `RedeemProcess._redeemStakeUsd()` internally call `UsdPool::subStableToken()`, then will revert,    which causes users to be unable to claim rewards, reduce positions, and redeem StakeUsd. Additionally, if no new funds are added by other users, the impact could last for over a week.
> According to [Sherlock’s rules](https://docs.sherlock.xyz/audits/judging/judging#v.-how-to-identify-a-medium-issue), “The issue causes locking of funds for users for more than a week,” it qualifies as High.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by ghost** [comment]
I understand the path which could lead to the scenario mentioned, my question is what exactly would be the impact? How did you come to conclusion that DOS could last for over a week not a day or month??
@0502lian @ZeroTrust01

**Comment by 0502lian** [comment]
> if no new funds are added by other users, the impact could last for over a week.

if no new funds are added by other users, the impact could last for over a week.  This mentions a possibility. Our job in auditing is to identify various issues that can affect the system. Regarding this issue, the larger the amount users want to reduce their positions and redeem, the higher the probability of Revert caused by calculation errors. The longer the time, the higher the probability of new funds coming in from users. In extreme cases, it could take a month, although the probability is very low. @Hash01011122

**Comment by mstpr** [comment]
Escalate

This should be invalid. `unsettledFee` is not liquid in the contract until it is settled. Including it in the accounting does not make sense because `unsettledFee` does not exist in the pools as a token. While accounting for it in the total value makes sense, both functions mentioned are strictly for the actual token balances, and `unsettledFee` does not exist in those balances.

**Comment by sherlock-admin3** [comment]
> Escalate
> 
> This should be invalid. `unsettledFee` is not liquid in the contract until it is settled. Including it in the accounting does not make sense because `unsettledFee` does not exist in the pools as a token. While accounting for it in the total value makes sense, both functions mentioned are strictly for the actual token balances, and `unsettledFee` does not exist in those balances.
> 
> 

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by WangSecurity** [comment]
I agree with @mstpr comment, but to make sure I totally understand the issue, what are the pre-requisites for this issue to occur? Just as soon as the users start using the protocol? @0502lian could you make a small scenario how this issue would occur with real numbers that would be used and the links to appropriate functions?

**Comment by 0502lian** [comment]
> I agree with @mstpr comment, but to make sure I totally understand the issue, what are the pre-requisites for this issue to occur? Just as soon as the users start using the protocol? @0502lian could you make a small scenario how this issue would occur with real numbers that would be used and the links to appropriate functions?


@mstpr Escalate both issue 236 and issue 237.
Issue 236 and issue 237 are somewhat related and comparable. I have pointed out the flaws in his views and partially answered your questions in the latest comments on issue 237.
Perhaps we should wait and see what he says.  @WangSecurity

**Comment by WangSecurity** [comment]
I agree this is a valid issue, but medium severity is more appropriate here. As I understand, there are specific conditions required for this to happen and it won't happen always with every user. Secondly, about the following:

> According to [Sherlock’s rules](https://docs.sherlock.xyz/audits/judging/judging#v.-how-to-identify-a-medium-issue), “The issue causes locking of funds for users for more than a week,” it qualifies as High

The rule doesn't say "The issue causes locking of funds for users for more than a week" is necessarily high severity and medium is more appropriate here.

Planning to reject both escalations and leave the issue as it is.

**Comment by WangSecurity** [comment]
Result:
Medium
Unique

**Comment by sherlock-admin3** [comment]
Escalations have been resolved successfully!

Escalation status:
- [0502lian](https://github.com/sherlock-audit/2024-05-elfi-protocol-judging/issues/236/#issuecomment-2210368272): rejected
- [mstpr](https://github.com/sherlock-audit/2024-05-elfi-protocol-judging/issues/236/#issuecomment-2210744169): rejected

**Comment by sherlock-admin2** [comment]
The Lead Senior Watson signed off on the fix.
